demos/common/processing/async_pipeline.cpp

```diff
@@ -0,0 +1,217 @@
+    return frameID;
+}
+
+int64_t PipelineBase::submitImage(cv::Mat img) {
+    auto request = requestsPool->getIdleRequest();
+    if (!request)
+        return -1;
+
+    MetaData* md;
+    model->preprocess(ImageInputData(img), request, md);
+
+    return submitRequest(request, std::shared_ptr<MetaData>(md));
+}
+
+std::unique_ptr<ResultBase> PipelineBase::getResult(bool shouldKeepOrder)
+{
+    auto infResult = PipelineBase::getInferenceResult(shouldKeepOrder);
+    if (infResult.IsEmpty()) {
+        return std::unique_ptr<ResultBase>();
+    }
+
+    auto result = model->postprocess(infResult);
+
+    *result.get() = static_cast<ResultBase&>(infResult);
+    return result;
+}
+
+InferenceResult PipelineBase::getInferenceResult(bool shouldKeepOrder)
+{
+    std::lock_guard<std::mutex> lock(mtx);
+
+    const auto& it = shouldKeepOrder ?
+        completedInferenceResults.find(outputFrameId) :
+        completedInferenceResults.begin();
+
+    if (it != completedInferenceResults.end())
+    {
+        auto retVal = std::move(it->second);
+        completedInferenceResults.erase(it);
+
+        outputFrameId = retVal.frameId;
+        outputFrameId++;
+        if (outputFrameId < 0)
+            outputFrameId = 0;
+
+        // Updating performance info
+        auto now = std::chrono::steady_clock::now();
+        auto latency = now - retVal.startTime;
+
+        auto oldLatency = perfInternals.latenciesMs[perfInternals.currentIndex];
+        auto oldRetrievalTimestamp = perfInternals.retrievalTimestamps[perfInternals.currentIndex];
+
+        perfInternals.latenciesMs[perfInternals.currentIndex] = std::chrono::duration_cast<std::chrono::milliseconds>(latency).count();
+        perfInternals.movingLatenciesSumMs += perfInternals.latenciesMs[perfInternals.currentIndex];
+        perfInternals.retrievalTimestamps[perfInternals.currentIndex] = now;
+        perfInternals.currentIndex = (perfInternals.currentIndex + 1) % MOVING_AVERAGE_SAMPLES;
+
+        perfInfo.latencySum += latency;
+        perfInfo.framesCount++;
+        perfInfo.FPS = perfInfo.framesCount*1000.0 / std::chrono::duration_cast<std::chrono::milliseconds>(now - perfInfo.startTime).count();
+
+        if (perfInfo.framesCount <= MOVING_AVERAGE_SAMPLES) {
+            // History buffer is not full in the beginning, so let's reuse global metrics so far
+            perfInfo.movingAverageLatencyMs = ((double)perfInternals.movingLatenciesSumMs) / perfInfo.framesCount;
+            perfInfo.movingAverageFPS = perfInfo.FPS;
+        }
+        else
+        {
+            // Now history buffer is full, we can use its data
+            perfInternals.movingLatenciesSumMs -= oldLatency;
+            perfInfo.movingAverageLatencyMs = ((double)perfInternals.movingLatenciesSumMs) / MOVING_AVERAGE_SAMPLES;
+            perfInfo.movingAverageFPS = MOVING_AVERAGE_SAMPLES * 1000.0 /
+                std::chrono::duration_cast<std::chrono::milliseconds>(now - oldRetrievalTimestamp).count();
```

Comment:
Reporting moving average leads to results which change every frame.

Reuse `PerformanceMetrics`. `PerformanceMetrics` is already declared as common. If you are not fine with `PerformanceMetrics`'s functionality you need to update it instead of reimplementing your own.

`PerformanceMetrics` doesn't report inference time and I don't think we need this as we can't report inference only time. The measurement includes waiting in IE's queue.

Reply:
PerformanceMetrics does not provide calculating moving average. I can update and reuse it, but it is separate task and should be done in separate PR.

Reply:
The reason for
> PerformanceMetrics does not provide calculating moving average.

is
> Reporting moving average leads to results which change every frame.

> I can update and reuse it, but it is separate task and should be done in separate PR.

Don't mess with metrics here at all then.

Reply:
not clear. It's better use moving average as it gives more smooth result. PerformanceMetrics cannot do it. As soon as it will be able to do it, I could reuse the code. But for now it cannot, why should we use it ?

Reply:
@fzhar reuse existing code as much as possible. I could say reporting moving average is separate task and can be done in separate PR. Just use reporting what we have now. Do not try to solve many different task simultaneously.

Reply:
done